pkg/controller/backupinfrastructure/backup_infrastructure_control.go

```diff
@@ -0,0 +1,338 @@
+	secrets            map[string]*corev1.Secret
+	imageVector        imagevector.ImageVector
+	identity           *gardenv1beta1.Gardener
+	config             *componentconfig.ControllerManagerConfiguration
+	recorder           record.EventRecorder
+	updater            UpdaterInterface
+}
+
+func (c *defaultControl) ReconcileBackupInfrastructure(obj *gardenv1beta1.BackupInfrastructure, key string) error {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return err
+	}
+
+	var (
+		backupInfrastructure       = obj.DeepCopy()
+		backupInfrastructureLogger = logger.NewFieldLogger(logger.Logger, "backupinfrastructure", fmt.Sprintf("%s/%s", backupInfrastructure.Namespace, backupInfrastructure.Name))
+		lastOperation              = backupInfrastructure.Status.LastOperation
+		operationType              = controllerutils.ComputeOperationType(lastOperation)
+	)
+
+	logger.Logger.Infof("[BACKUPINFRASTRUCTURE RECONCILE] %s", key)
+	backupInfrastructureJSON, _ := json.Marshal(backupInfrastructure)
+	backupInfrastructureLogger.Debugf(string(backupInfrastructureJSON))
+
+	op, err := operation.NewWithBackupInfrastructure(backupInfrastructure, backupInfrastructureLogger, c.k8sGardenClient, c.k8sGardenInformers, c.identity, c.secrets, c.imageVector, c.config)
+	if err != nil {
+		backupInfrastructureLogger.Errorf("Could not initialize a new operation: %s", err.Error())
+		return err
+	}
+
+	// The deletionTimestamp labels a BackupInfrastructure as intended to get deleted. Before deletion,
+	// it has to be ensured that no infrastructure resources are depending on the BackupInfrastructure anymore.
+	// When this happens the controller will remove the finalizer from the BackupInfrastructure so that it can be garbage collected.
+	if backupInfrastructure.DeletionTimestamp != nil {
+		gracePeriod := time.Hour * 24 * time.Duration(*c.config.Controllers.BackupInfrastructure.DeletionGracePeriodDays)
+		if time.Now().Sub(backupInfrastructure.DeletionTimestamp.Time) > gracePeriod {
+			if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateProcessing, gardenv1beta1.ShootLastOperationTypeDelete, "Deletion of Backup Infrastructure in progress.", 1, nil); updateErr != nil {
+				backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after deletion start: %+v", updateErr)
+				return updateErr
+			}
+
+			if deleteErr := c.deleteBackupInfrastructure(op); deleteErr != nil {
+				c.recorder.Eventf(backupInfrastructure, corev1.EventTypeWarning, gardenv1beta1.EventDeleteError, "%s", deleteErr.Description)
+				if updateErr := c.updateBackupInfrastructureStatus(op, gardenv1beta1.ShootLastOperationStateError, gardenv1beta1.ShootLastOperationTypeDelete, deleteErr.Description+" Operation will be retried.", 1, deleteErr); updateErr != nil {
+					backupInfrastructureLogger.Errorf("Could not update the BackupInfrastructure status after deletion error: %+v", updateErr)
+					return updateErr
```

Comment:
This i just followed from shoot controller. But i thinks its fine. We don't need to add to return error message. Anyways we can find delete error description with `kubectl describe`.  And we are not logging error message return from `ReconcileBackupInfrastructure`.  If `updateErr` would have been caused because of `deleteErr` then it would have make sense to concatenate it.

Reply:
Hm, ok, agreed.